tpm2/platform_legacy.go

```diff
@@ -0,0 +1,132 @@
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package tpm2
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+
+	"golang.org/x/xerrors"
+
+	"github.com/snapcore/secboot"
+)
+
+const legacyPlatformName = "tpm2-legacy"
+
+type legacyPlatformKeyDataHandler struct{}
+
+func (h *legacyPlatformKeyDataHandler) RecoverKeys(data *secboot.PlatformKeyData) (secboot.KeyPayload, error) {
+	tpm, err := ConnectToTPM()
+	switch {
+	case err == ErrNoTPM2Device:
+		return nil, &secboot.PlatformKeyRecoveryError{
+			Type: secboot.PlatformKeyRecoveryErrorUnavailable,
+			Err:  err}
+	case err != nil:
+		return nil, xerrors.Errorf("cannot connect to TPM: %w", err)
+	}
+	defer tpm.Close()
+
+	var handle []byte
+	if err := json.Unmarshal(data.Handle, &handle); err != nil {
+		return nil, &secboot.PlatformKeyRecoveryError{
+			Type: secboot.PlatformKeyRecoveryErrorInvalidData,
+			Err:  err}
+	}
+
+	k, err := ReadSealedKeyObject(bytes.NewReader(handle))
+	if err != nil {
+		var e InvalidKeyDataError
+		if xerrors.As(err, &e) {
+			return nil, &secboot.PlatformKeyRecoveryError{
+				Type: secboot.PlatformKeyRecoveryErrorInvalidData,
+				Err:  err}
+		}
+		return nil, xerrors.Errorf("cannot read key object: %w", err)
+	}
+
+	key, authKey, err := k.UnsealFromTPM(tpm)
+	if err != nil {
+		var e InvalidKeyDataError
+		switch {
+		case xerrors.As(err, &e):
+			return nil, &secboot.PlatformKeyRecoveryError{
+				Type: secboot.PlatformKeyRecoveryErrorInvalidData,
+				Err:  errors.New(e.msg)}
+		case err == ErrTPMProvisioning:
+			return nil, &secboot.PlatformKeyRecoveryError{
+				Type: secboot.PlatformKeyRecoveryErrorUninitialized,
+				Err:  err}
+		case err == ErrTPMLockout:
+			return nil, &secboot.PlatformKeyRecoveryError{
+				Type: secboot.PlatformKeyRecoveryErrorUnavailable,
+				Err:  err}
+		}
+		return nil, xerrors.Errorf("cannot unseal key: %w", err)
+	}
+
+	return secboot.MarshalKeys(key, secboot.AuxiliaryKey(authKey)), nil
+}
+
+// NewKeyDataFromSealedKeyObjectFile creates a secboot.KeyData for the TPM
+// sealed key object at the supplied path, in order to enable keys to be
+// recovered from the TPM sealed key object using the secboot.KeyData API.
+//
+// Note that the returned KeyData does not support the snap model authorization
+// API, and consumers of this function should not attempt to use this API.
```

Comment:
this needs to be known implicitly ATM?  a modchecker will still be returned?

Reply:
That's right - if this API is used to load the key data, then IsModelAuthorized shouldn't be used because it will return false.

When we have new-style TPM objects, I can add a function to determine which API to use to load a file.